### PR TITLE
feat: add strategy

### PR DESCRIPTION
**Pull request by steffenix**

### CODE REVIEW

**Review by mushroomsforest — COMMENTED** [review]
looks good to me in general

File: contracts/Strategy.sol [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+        }
+
+        uint256 _wantProfit;
+        IUniswapV2Router02 uniswapV2Router = IUniswapV2Router02(unirouter);
+        uniswapV2Router.swapExactTokensForTokens(
+            mmBalance,
+            uint256(1),
+            pathToSwap,
+            address(this),
+            now + 180
```
[the remark this answers is not in the document]
Comment (steffenix) [reply contracts/Strategy.sol:316]: Yes I can remove the + 180.

File: contracts/Strategy.sol [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+            _loss = _reportLoss;
+        }
+
+        // Claim profit
+        _profit = 0;
+        uint256 total = estimatedTotalAssets();
+
+        if (total > params.totalDebt) {
+            _profit = total.sub(params.totalDebt);
+            liquidatePosition(_profit);
```
Comment (Grandthrax) [inline contracts/Strategy.sol:115]: Combine with the liquidate position above so as to save gas. They are expensive operations so should be done just once

File: contracts/Strategy.sol [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+        }
+
+        if (params.totalDebt > total) {
```
Comment (Grandthrax) [inline contracts/Strategy.sol:118]: can use else

File: contracts/Strategy.sol [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+        uint256 _want = want.balanceOf(address(this));
+
+        if (_want > _debtOutstanding) {
+            _want = _want.sub(_debtOutstanding);
+
+            MMVault(mmVault).deposit(_want);
+
+            _after = IERC20(mmVault).balanceOf(address(this));
+            require(_after > _before, "!mismatchDepositIntoMushrooms");
+        } else if (_debtOutstanding > _want) {
```
Comment (Grandthrax) [inline contracts/Strategy.sol:146]: Don't need this. Can still do the (after > 0) if statement

File: contracts/Strategy.sol [hunk 5]
```diff
@@ -0,0 +1,382 @@
...
+            return;
+        }
+
+        emit DepoistedOnMMFarmingPool(_after);
```
Comment (Grandthrax) [inline contracts/Strategy.sol:150]: do you mean after.sub(before)?

File: contracts/Strategy.sol [hunk 6]
```diff
@@ -0,0 +1,382 @@
...
+        (uint256 amountA, uint256 amountB, uint256 LP) =
+            uniswapV2Router.addLiquidity(
+                ust,
+                mir,
+                ustAmount - amountToSwap,
+                amounts[1],
+                1,
+                1,
+                address(this),
+                now + 60
```
Comment (Grandthrax) [inline contracts/Strategy.sol:364]: now is fine

File: contracts/Strategy.sol [hunk 7]
```diff
@@ -0,0 +1,382 @@
...
+                        "!insufficientMTokenInMasterChef"
+                    );
+                    MMFarmingPool(mmFarmingPool).withdraw(
+                        mmFarmingPoolId,
+                        _mvGap
+                    );
+                }
+                MMVault(mmVault).withdraw(_mShare);
+                uint256 _after = IERC20(want).balanceOf(address(this));
+                require(_after > _before, "!mismatchMushroomsVaultWithdraw");
```
Comment (Grandthrax) [inline contracts/Strategy.sol:221]: does this ever trigger? what situation is it for?

File: contracts/Strategy.sol [hunk 8]
```diff
@@ -0,0 +1,382 @@
...
+                    _after,
+                    _amountNeeded > _after ? _amountNeeded.sub(_after) : 0
+                );
+            } else {
+                return (_amountNeeded, _loss);
+            }
+        }
+    }
+
+    function harvestTrigger(uint256 callCost)
```
Comment (Grandthrax) [inline contracts/Strategy.sol:233]: call cost is in eth. you need to convert it to want

Comment (Grandthrax) [reply contracts/Strategy.sol:233]: What is different in this method than in base strategy? If you are making no changes then the method should look like this:

```
function harvestTrigger(uint256 callCost) public view override returns (bool) {
        uint256 wantCallCost = _callCostToWant(callCost);
        return super.harvestTrigger(wantCallCost);
    }
```

File: contracts/Strategy.sol [hunk 9]
```diff
@@ -0,0 +1,382 @@
...
+
+    address public constant mm =
+        address(0xa283aA7CfBB27EF0cfBcb2493dD9F4330E0fd304);
+    address public constant mir =
+        address(0x09a3EcAFa817268f77BE1283176B946C4ff2E608);
+    address public constant ust =
+        address(0xa47c8bf37f92aBed4A126BDA807A7b7498661acD);
+
+    uint256 public minMMToSwap = 10; // min $MM to swap during adjustPosition()
+    address[] private pathToSwap;
```
Comment (Grandthrax) [inline contracts/Strategy.sol:53]: Personally I would put in the normal route here. As you have it it could be forgotten during setup. Don't put in constructor as having arrays in constructor can mess up verification on etherscan

File: tests/test_operation.py [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+
+
+def test_triggers(gov, vault, strategy, token, amount, weth, weth_amout):
+    # Deposit to the vault and harvest
+    token.approve(vault.address, amount, {"from": gov})
+    vault.deposit(amount, {"from": gov})
+    vault.updateStrategyDebtRatio(strategy.address, 5_000, {"from": gov})
+    strategy.harvest()
+
+    strategy.harvestTrigger(0)
```
Comment (Grandthrax) [inline tests/test_operation.py:118]: 0 is not a useful input in trigger. it always returns true

File: tests/test_operation.py [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+    assert pytest.approx(strategy.estimatedTotalAssets(), rel=1e-5) == amount
+
+    # set emergency and exit
+    strategy.setEmergencyExit()
+    strategy.harvest()
+
+    assert strategy.estimatedTotalAssets() < amount
+
+
+def test_profitable_harvest(
```
Comment (Grandthrax) [inline tests/test_operation.py:39]: I would like an apr test please. Show us how much it is going to earn.

File: tests/test_operation.py [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+
+def test_triggers(gov, vault, strategy, token, amount, weth, weth_amout):
+    # Deposit to the vault and harvest
+    token.approve(vault.address, amount, {"from": gov})
+    vault.deposit(amount, {"from": gov})
+    vault.updateStrategyDebtRatio(strategy.address, 5_000, {"from": gov})
+    strategy.harvest()
+
+    strategy.harvestTrigger(0)
+    strategy.tendTrigger(0)
```
Comment (Grandthrax) [inline tests/test_operation.py:119]: You don't use tend trigger as far as i can see?

**Review by Grandthrax — COMMENTED** [review]
Looks good. Very neat code. Nice one. Some improvements but generally good

File: contracts/Strategy.sol (hunk 9 above)
Comment (steffenix) [reply contracts/Strategy.sol:53]: I can't find the right syntax to add values to an array on here.

File: tests/test_operation.py (hunk 11 above)
Comment (steffenix) [reply tests/test_operation.py:39]: APY: 98.7258129416045 %

File: contracts/Strategy.sol [hunk 13]
```diff
@@ -0,0 +1,408 @@
...
+// Import interfaces for many popular DeFi projects, or add your own!
+//import "../interfaces/<protocol>/<Interface>.sol";
+
+contract Strategy is BaseStrategy {
+    using SafeERC20 for IERC20;
+    using Address for address;
+    using SafeMath for uint256;
+
+    event DepoistedOnMMFarmingPool(uint256 amount);
+    event ProfitDetails(uint256 appreciation, uint256 sellingMushroom);
```
[the remark this answers is not in the document]
Comment (steffenix) [reply contracts/Strategy.sol:34]: That was great for debug, I can remove it.

File: contracts/Strategy.sol [hunk 14]
```diff
@@ -0,0 +1,408 @@
...
+        uint256 _after = _before;
+        uint256 _want = want.balanceOf(address(this));
+
+        if (_want > _debtOutstanding) {
+            _want = _want.sub(_debtOutstanding);
+
+            MMVault(mmVault).deposit(_want);
+
+            _after = IERC20(mmVault).balanceOf(address(this));
+            require(_after > _before, "!mismatchDepositIntoMushrooms");
```
[the remark this answers is not in the document]
Comment (steffenix) [reply contracts/Strategy.sol:144]: That's a safety measure, that was part of the strategy this strategy has been built on top of.

File: contracts/Strategy.sol [hunk 15]
```diff
@@ -0,0 +1,408 @@
...
+
+    function liquidatePosition(uint256 _amountNeeded)
+        internal
+        override
+        returns (uint256 _liquidatedAmount, uint256 _loss)
+    {
+        bool liquidateAll =
+            _amountNeeded >= estimatedTotalAssets() ? true : false;
+
+        if (liquidateAll) {
```
[the remark this answers is not in the document]
Comment (steffenix) [reply contracts/Strategy.sol]: I have refactored the `liquidatePosition`:

https://github.com/steffenix/yearn-strategy-mir-ust/pull/1/commits/e68d0b2f6edce5c4f880dca88f49d5eb62d45887#diff-3e804d467472e49fe8afea1899d96f26655f3ecd641555a801abcea4f12af598R162

File: contracts/Strategy.sol [hunk 16]
```diff
@@ -0,0 +1,408 @@
...
+            uint256 _debtPayment
+        )
+    {
+        StrategyParams memory params = vault.strategies(address(this));
+
+        _profit = 0;
+        uint256 total = estimatedTotalAssets();
+
+        if (_debtOutstanding > 0 || total > params.totalDebt) {
+            if (total > params.totalDebt) _profit = total.sub(params.totalDebt);
```
[the remark this answers is not in the document]
Comment (steffenix) [reply contracts/Strategy.sol:103]: The original code was:

```
    function prepareReturn(uint256 _debtOutstanding)
        internal
        override
        returns (
            uint256 _profit,
            uint256 _loss,
            uint256 _debtPayment
        )
    {
        StrategyParams memory params = vault.strategies(address(this));

        // Pay debt if any
        if (_debtOutstanding > 0) {
            (uint256 _amountFreed, uint256 _reportLoss) =
                liquidatePosition(_debtOutstanding);
            _debtPayment = _amountFreed > _debtOutstanding
                ? _debtOutstanding
                : _amountFreed;
            _loss = _reportLoss;
        }

        // Claim profit
        _profit = 0;
        uint256 total = estimatedTotalAssets();

        if (total > params.totalDebt) {
            _profit = total.sub(params.totalDebt);
            liquidatePosition(_profit);
        }

        uint256 claimed = claimMM();
        emit ProfitDetails(_profit, claimed);

        _profit += claimed;
        return (_profit, _loss, _debtPayment);
    }
```

but from a suggestion from @Grandthrax we should group `liquidatePosition`.
Can you suggest a way to do it?

[the remark this answers is not in the document]
Comment (steffenix) [reply contracts/Strategy.sol]: I have reverted to this code.

It's not possible to `estimatedTotalAssets` before liquidating potential debt.

File: contracts/Strategy.sol [hunk 17]
```diff
@@ -0,0 +1,389 @@
...
+    }
+
+    function convertEthToWant(uint256 amount) public view returns (uint256) {
+        address[] memory path = new address[](3);
+        path[0] = weth;
+        path[1] = usdt;
+        path[2] = ust;
+
+        uint256 ustAmount =
+            IUniswapV2Router02(unirouter).getAmountsOut(amount, path)[2];
```
Comment (poolpitako) [inline contracts/Strategy.sol:270]: you need to add something like
```
if (amount == 0)
  return 0;
```

This call will fail with amount == 0.

File: contracts/Strategy.sol [hunk 18]
```diff
@@ -0,0 +1,389 @@
...
+        uint256 total = estimatedTotalAssets();
+
+        if (total > params.totalDebt) {
+            _profit = total.sub(params.totalDebt);
+            (_profit, ) = liquidatePosition(_profit);
+        }
+
+        uint256 claimed = claimMM();
+
+        _profit += claimed;
```
Comment (poolpitako) [inline contracts/Strategy.sol:119]: if liquidatePosition in line 103 has a loss, and you can claim MM in line 117 you would be reporting loss and profit at the same time.

File: contracts/Strategy.sol [hunk 19]
```diff
@@ -0,0 +1,389 @@
...
+
+        if (_want > _debtOutstanding) {
+            _want = _want.sub(_debtOutstanding);
+
+            MMVault(mmVault).deposit(_want);
+
+            _after = IERC20(mmVault).balanceOf(address(this));
+        }
+
+        if (_after > 0) {
```
Comment (poolpitako) [inline contracts/Strategy.sol:141]: Is this needed? can't you remove `_before` and `_after` logic and just do:
```
uint256 mmBalance = IERC20(mmVault).balanceOf(address(this));
if (mmBalance > 0) {
  MMFarmingPool(mmFarmingPool).deposit(mmFarmingPoolId, _after);
}
```

My understanding is: if you have debt, don't deposit want, but if you have mmVault, you should always deposit if there is balance left.

File: contracts/Strategy.sol [hunk 20]
```diff
@@ -0,0 +1,389 @@
...
+        }
+
+        uint256 total = estimatedTotalAssets();
+
+        if (total > params.totalDebt) {
+            _profit = total.sub(params.totalDebt);
+            (_profit, ) = liquidatePosition(_profit);
+        }
+
+        uint256 claimed = claimMM();
```
Comment (poolpitako) [inline contracts/Strategy.sol:117]: why don't you use balance instead of trusting function returns?

File: contracts/Strategy.sol (hunk 8 above)
Comment (poolpitako) [reply contracts/Strategy.sol:210]: status of this?

File: contracts/Strategy.sol [hunk 21]
```diff
@@ -0,0 +1,389 @@
...
+        internal
+        view
+        returns (uint256)
+    {
+        uint256 _mTokenTotal = IERC20(mmVault).totalSupply();
+        if (_mTokenTotal == 0) {
+            return 0;
+        }
+        uint256 _wantInVault = MMVault(mmVault).balance();
+        return (_wantInVault.mul(_shares)).div(_mTokenTotal).mul(998).div(1000);
```
Comment (poolpitako) [inline contracts/Strategy.sol:321]: why wouldn't you use `getAmountsOut()` instead of doing the math by hand?

File: contracts/Strategy.sol (hunk 19 above)
Comment (steffenix) [reply contracts/Strategy.sol:141]: Good point.

File: contracts/Strategy.sol (hunk 20 above)
Comment (steffenix) [reply contracts/Strategy.sol:117]: replaced with:
_profit =  IERC20(want).balanceOf(address(this));

File: contracts/Strategy.sol (hunk 21 above)
Comment (steffenix) [reply contracts/Strategy.sol:321]: That looks like the getAmountOut from uniswap but that is used to take into account the 0.2% fees from the MM vault.

File: contracts/Strategy.sol [hunk 22]
```diff
@@ -0,0 +1,390 @@
...
+            );
+        }
+
+        return (_liquidatedAmount, _loss);
+    }
+
+    function harvestTrigger(uint256 callCost)
+        public
+        view
+        override
```
Comment (poolpitako) [inline contracts/Strategy.sol:209]: @steffenix can you answer the status of this change? are you copy/pasting basestrategy's method?

File: contracts/Strategy.sol [hunk 23]
```diff
@@ -0,0 +1,390 @@
...
+            MMVault(mmVault).deposit(_want);
+        }
+
+        uint256 mmBalance = IERC20(mmVault).balanceOf(address(this));
+        if (mmBalance > 0) {
+            MMFarmingPool(mmFarmingPool).deposit(mmFarmingPoolId, mmBalance);
+        }
+    }
+
+    function claimMM() internal returns (uint256) {
```
Comment (poolpitako) [inline contracts/Strategy.sol:142]: You are not longer calling `claimMM`?

File: contracts/Strategy.sol [hunk 24]
```diff
@@ -0,0 +1,390 @@
...
+        override
+        returns (uint256, uint256)
+    {
+        uint256 _liquidatedAmount;
+        uint256 _loss;
+        uint256 totalAssets = estimatedTotalAssets();
+
+        if (_amountNeeded > totalAssets) {
+            _liquidatedAmount = totalAssets;
+            _loss = _amountNeeded.sub(totalAssets);
```
Comment (poolpitako) [inline contracts/Strategy.sol:165]: you should calculate the loss at the end of the process and not basing on the `estimatedTotalAssets()`. You might withdraw 0.

File: contracts/Strategy.sol [hunk 25]
```diff
@@ -0,0 +1,390 @@
...
+            _debtPayment = _amountFreed > _debtOutstanding
+                ? _debtOutstanding
+                : _amountFreed;
+            _loss = _reportLoss;
+        } else {
+            uint256 total = estimatedTotalAssets();
+
+            if (total > params.totalDebt) {
+                _profit = total.sub(params.totalDebt);
+                liquidatePosition(_profit);
```
Comment (poolpitako) [inline contracts/Strategy.sol:113]: `_profit` value is overriden in line 116.
I this this should be:
```
liquidatePosition(total.sub(params.totalDebt));
```

File: contracts/Strategy.sol (hunk 22 above)
Comment (steffenix) [reply contracts/Strategy.sol:209]: The changes are on line 241:
```
            callCost = convertEthToWant(callCost);
```

Comment (poolpitako) [reply contracts/Strategy.sol:209]: Why don't you do:
```
function harvestTrigger(uint256 callCost) public view override returns (bool) {
        uint256 wantCallCost = _callCostToWant(callCost);
        return super.harvestTrigger(wantCallCost);
}
```

File: contracts/Strategy.sol (hunk 23 above)
Comment (steffenix) [reply contracts/Strategy.sol:142]: mistake.

File: contracts/Strategy.sol (hunk 22 above)
Comment (steffenix) [reply contracts/Strategy.sol:209]: The changes are on line 241:
```
            callCost = convertEthToWant(callCost);
```

File: contracts/Strategy.sol (hunk 25 above)
Comment (steffenix) [reply contracts/Strategy.sol:113]: ok

File: contracts/Strategy.sol (hunk 24 above)
Comment (steffenix) [reply contracts/Strategy.sol:165]: I do not understand how I should do it.

Comment (steffenix) [reply contracts/Strategy.sol:165]: @poolpitako Can you help with that?

Comment (Grandthrax) [reply contracts/Strategy.sol:165]: I dont understand either?

**Comment by Grandthrax** [comment]
https://github.com/steffenix/yearn-strategy-mir-ust/blob/master/contracts/Strategy.sol#L109

Is there ever a time when there is want in balance that is not profit?

**Comment by Grandthrax** [comment]
https://github.com/steffenix/yearn-strategy-mir-ust/blob/master/contracts/Strategy.sol#L128

Not used?

**Comment by steffenix** [comment]
> https://github.com/steffenix/yearn-strategy-mir-ust/blob/master/contracts/Strategy.sol#L109
> 
> Is there ever a time when there is want in balance that is not profit?

No, 
wants are coming from withdrawing from the underlying vault and selling mushroom.
After the profit is reported back to the vault and reinvested into the vault if leftover.
